src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Schemas/JSON/template.json

```diff
@@ -1170,21 +1170,98 @@
       "type": "object",
       "additionalProperties": {
         "type": "object",
         "required": [ "type" ],
         "properties": {
           "type": {
             "type": "string",
-            "description": "constraint type"
+            "description": "Constraint type."
           },
           "args": {
-            "type": ["array", "string", "object"],
-            "description": "constraint arguments"
+            "description": "Constraints arguments."
           }
-        }
+        },
+        "oneOf": [
+          {
+            "not": {
+              "required": [ "type" ],
+              "properties": {
+                "type": {
+                  "enum": [
+                    "os",
+                    "host"
+                  ]
+                }
+              }
+            },
+            "properties": {
+              "args": {
+                "type": [ "array", "object", "string" ]
+              }
+            }
+          },
+          {
+            "description": "Operating system constraint. Allows to restrict the template to be run only on specific operating systems.",
+            "properties": {
+              "type": {
+                "enum": [ "os" ]
+              },
+              "args": {
+                "oneOf": [
+                  {
+                    "type": "string",
+                    "enum": [ "Windows", "Unix", "OSX" ]
+                  },
+                  {
+                    "type": "array",
+                    "items": {
+                      "type": "string",
+                      "enum": [ "Windows", "Unix", "OSX" ]
+                    }
+                  }
+                ]
+              }
+            }
+          },
+          {
+            "description": "Host constraint. Allows to restrict the template to be run only on specific applications using template engine.",
+            "properties": {
+              "type": {
+                "enum": [ "host" ]
+              },
+              "args": {
+                "type": "array",
+                "items": {
+                  "type": "object",
+                  "required": [ "hostname" ],
+                  "properties": {
+                    "hostname": {
+                      "type": "string",
+                      "description": "The identifier of supported host. Commonly used identifiers are: 'dotnetcli'' - .NET SDK, 'vs' - Visual Studio, 'vs-mac' - Visual Studio for Mac, 'ide' - Visual Studio (any platform).",
+                      "oneOf": [
+                        {
+                          "not": {
+                            "enum": [ "dotnetcli", "vs", "vs-mac", "ide" ]
+                          }
+                        },
+                        {
+                          "enum": [ "dotnetcli", "vs", "vs-mac", "ide" ]
+                        }
+                      ]
+                    },
+                    "version": {
```

Comment:
Do we want deeper validation for the versions?
E.g. something like: https://github.com/SchemaStore/schemastore/blob/master/src/schemas/json/global.json#L9

Reply:
On a second thought - it would probably get unreasonably complicated for all the syntaxes of expressing supported versions that we support (exact, range, float range ...)